gtk4/src/subclass/tree_model.rs

```diff
@@ -0,0 +1,503 @@
+            let iface = glib::gobject_ffi::g_type_default_interface_ref(type_)
+                as *mut ffi::GtkTreeModelIface;
+            assert!(!iface.is_null());
+
+            if let Some(f) = (*iface).row_deleted.as_ref() {
+                f(
+                    tree_model.unsafe_cast_ref::<TreeModel>().to_glib_none().0,
+                    path.to_glib_none().0 as *mut _,
+                );
+            }
+
+            glib::gobject_ffi::g_type_default_interface_unref(iface as glib::ffi::gpointer);
+        }
+    }
+
+    fn get_flags(&self, tree_model: &Self::Type) -> TreeModelFlags {
+        unsafe {
+            let type_ = ffi::gtk_print_operation_preview_get_type();
+            let iface = glib::gobject_ffi::g_type_default_interface_ref(type_)
+                as *mut ffi::GtkTreeModelIface;
+            assert!(!iface.is_null());
+
+            let f = (*iface).get_flags.as_ref().unwrap();
+            let ret = f(tree_model.unsafe_cast_ref::<TreeModel>().to_glib_none().0);
+
+            glib::gobject_ffi::g_type_default_interface_unref(iface as glib::ffi::gpointer);
+            from_glib(ret)
+        }
+    }
+
+    fn get_n_columns(&self, tree_model: &Self::Type) -> i32;
+    fn get_column_type(&self, tree_model: &Self::Type, index: i32) -> glib::Type;
+    fn get_iter(&self, tree_model: &Self::Type, path: &TreePath) -> Option<TreeIter>;
+    fn get_path(&self, tree_model: &Self::Type, iter: &TreeIter) -> TreePath;
+    fn get_value(&self, tree_model: &Self::Type, iter: &TreeIter, index: i32) -> Value;
+    fn iter_next(&self, tree_model: &Self::Type) -> Option<TreeIter>;
+
+    fn iter_previous(&self, tree_model: &Self::Type) -> Option<TreeIter> {
+        unsafe {
+            let type_ = ffi::gtk_print_operation_preview_get_type();
+            let iface = glib::gobject_ffi::g_type_default_interface_ref(type_)
+                as *mut ffi::GtkTreeModelIface;
+            assert!(!iface.is_null());
+            let iter = std::ptr::null_mut();
+            let f = (*iface).iter_previous.as_ref().unwrap();
+            let ret: bool = from_glib(f(
+                tree_model.unsafe_cast_ref::<TreeModel>().to_glib_none().0,
+                iter,
+            ));
+
+            glib::gobject_ffi::g_type_default_interface_unref(iface as glib::ffi::gpointer);
+            if ret {
+                let iter = from_glib_none(iter);
+                Some(iter)
+            } else {
+                None
+            }
+        }
+    }
+
+    fn iter_has_child(&self, tree_model: &Self::Type, iter: &TreeIter) -> bool;
+    fn iter_n_children(&self, tree_model: &Self::Type, iter: Option<&TreeIter>) -> i32;
+    fn iter_nth_child(
+        &self,
+        tree_model: &Self::Type,
+        parent: Option<&TreeIter>,
+        index: i32,
+    ) -> Option<TreeIter>;
+    fn iter_parent(&self, tree_model: &Self::Type, child: &TreeIter) -> Option<TreeIter>;
+
+    fn ref_node(&self, tree_model: &Self::Type, iter: &TreeIter) {
```

Comment:
Can we safely (or at all) implement this and the unref function? How is that supposed to work?

Reply:
The API docs mentions that the functions handles a ref counter of the currently displayed nodes so that they are cached. Here's an example from gtk of how such methods looks like 
```c
static void
gtk_tree_model_ref_count_ref_node (GtkTreeModel *model,
                                   GtkTreeIter  *iter)
{
  NodeInfo *info;
  GtkTreeModelRefCount *ref_model = GTK_TREE_MODEL_REF_COUNT (model);

  info = g_hash_table_lookup (ref_model->priv->node_hash, iter->user_data);
  if (!info)
    {
      info = node_info_new ();

      g_hash_table_insert (ref_model->priv->node_hash, iter->user_data, info);
    }

  info->ref_count++;
}

static void
gtk_tree_model_ref_count_unref_node (GtkTreeModel *model,
                                     GtkTreeIter  *iter)
{
  NodeInfo *info;
  GtkTreeModelRefCount *ref_model = GTK_TREE_MODEL_REF_COUNT (model);

  info = g_hash_table_lookup (ref_model->priv->node_hash, iter->user_data);
  g_assert (info != NULL);
  g_assert (info->ref_count > 0);

  info->ref_count--;
}
```
But I'm not sure where that ref counting is used, as in where the caching happens. I guess we can mark them as unsafe and leave figuring out how this part of the api is supposed to work to someone that actually uses GtkTreeModel

Reply:
That would mean making the trait unsafe, not the function.